### DISCUSSION TITLE
Can I use string literals for DB url in schema.prisma?

### DISCUSSION THREAD

**Discussion by kennyma603** [prisma/prisma#12336] [Q&A] [answered]
For some reason our server infrastructure doesnt allow us to define new env.
So i cant define a new env for DatabaseURL and use it in schema.prisma file. 

however, we have the following in envs already.

POSTGRESQL_USERNAME=postgres
POSTGRESQL_USER_PASSWORD=pw
POSTGRESQL_ENDPOINT=localhost

in schema.prisma, i tired combine the envs together as url,

datasource db {
  provider = "postgresql"
  url      = `postgresql://env("POSTGRESQL_USERNAME"):$env("POSTGRESQL_USER_PASSWORD")@env("POSTGRESQL_ENDPOINT"):5432/postgres?schema=public`
}

note: i guess url has to set here in order for `prisma migrate dev` to work.

but apparently it doesn't work. If there a way to get it to work? or out of luck?

**Comment by janpio** [comment, marked as answer]
You might be able to achieve this via the `.env` file might be able to contruct new env variables from existing ones: https://www.prisma.io/docs/guides/development-environment/environment-variables#expanding-variables

Alternatively you might be able to set the `url` in code: https://www.prisma.io/docs/reference/api-reference/prisma-client-reference#datasources

**Reply by kennyma603** [reply]
Thanks for your reply @janpio .

I could change my .env locally the way i wanted, however, the envs for prod are out of my control, i can only access them but cannot create one.

For alternative solution, I have tried this: https://github.com/prisma/prisma/discussions/3561#discussioncomment-1551839
it requires setting db url to "" in schema.prisma.
however, this would fail if i do 'npx prisma migrate dev', because the url is empty there.

**Reply by janpio** [reply]
> I could change my .env locally the way i wanted, however, the envs for prod are out of my control, i can only access them but cannot create one.

This is a file that exists in your project root. Or will that also not work in your environment?



> For alternative solution, I have tried this: [#3561 (comment)](https://github.com/prisma/prisma/discussions/3561#discussioncomment-1551839)
> it requires setting db url to "" in schema.prisma.

For command that need a valid `url` value, you can just set the value in your schema file to a dummy value that is not going to be used. (Effectively only the protocal and something URL like is required).

> however, this would fail if i do 'npx prisma migrate dev', because the url is empty there.

This is unfortunately still true - outside of Prisma Client, you need to either hardcode the value in your schema file or use an environment variable that has the full connection string.

**Reply by kennyma603** [reply]
re `This is a file that exists in your project root. Or will that also not work in your environment?`
so we are using node, and we have 3 different envs, local, dev, prod. changes in local would go thru pipeline to get promoted to dev and prod.

for local envs, sure .env file is stored locally and i can add new variable with full db url . and do npx prisma migrate dev no problem.

the envs for dev and prod environments are set in a place that i have no control off. so i cant add a new variable with full db url. so in the pipeline when it tries to do 'npx prisma migrate deploy' wont work as I dont have the full db url

**Reply by janpio** [reply]
Please have a look at the link I included. This defines how you can use the `.env` file to combine existing env vars (the ones you mentioned before) into a new env var (that would then include the full connection string).

Can you run a script that sets an env variabel for the runtime of that script? How about the command line `set FOO=bar`?

**Reply by mbsanchez01** [reply]
I think that this is an important improvement for Prisma, they should be able to accept building variables in the schema using different env variables, because, the problem is that in local you can do whatever you want, but in prod, it is not the same. Most of the time, the developers don't have control over the CI/CD, because this is controlled by DevOps and they have their standards and specifications. 

Setting the DATABASE_URL locally could be an option, but in production you normally don't use env files, for instance we use task definitions in AWS, or the key storage like in Azure.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
